File: dart/test/unicode_rule_label_routes_test.dart
// FUTURE-PARITY-BACKLOG.10.5.0.2.1 — native Dart label parsing/validation.

import 'dart:convert';

import 'package:linkedspec_dart/linkedspec_dart.dart';
import 'package:test/test.dart';

void main() {
  const decomposedTop = 'To\u{0308}p';

  test(
    'scanner parses every declaration and edge form with exact identity',
    () {
      final action = parseSpec('''
Töp::AND
 -> $decomposedTop | Δοκιμή { return(entry_text()) }

$decomposedTop:
 /x/

Δοκιμή:
 /x/
''');
      validateSpec(action);
      expect(action.rules.map((rule) => rule.header.label), [
        'Töp',
        decomposedTop,
        'Δοκιμή',
      ]);
      final actionEdge =
          action.rules.first.body.single.kind as ActionEdgeBodyElementKind;
      expect(actionEdge.targets.map((target) => target.label), [
        decomposedTop,
        'Δοκιμή',
      ]);

      final blind = parseSpec('''
規則::AND
 => 𐐀Rule

𐐀Rule:
 /x/
''');
      validateSpec(blind);
      final blindEdge =
          blind.rules.first.body.single.kind as BlindEdgeBodyElementKind;
      expect(blindEdge.target, '𐐀Rule');

      final bare = parseSpec('''
9_root::OR
 A·B | _ { return(entry_text()) }

A·B:
 /x/

_:
 /x/
''');
      validateSpec(bare);
      final bareEdge =
          bare.rules.first.body.single.kind as BareEdgeBodyElementKind;
      expect(bareEdge.targets.map((target) => target.label), ['A·B', '_']);
    },
  );

  test('invalid suffixes never become partial action blind or bare edges', () {
    for (final invalid in [
      '',
      'Top-Rule',
      'Top Rule',
      'Top😀',
      'Top:',
      'Top/Rule',
      r'$Top',
    ]) {
      for (final arrow in ['->', '=>']) {
        final spec = parseSpec('Top::\n $arrow $invalid\n');
        expect(
          spec.rules.single.body.single.kind,
          isA<RawBodyElementKind>(),
          reason: '$arrow $invalid',
        );
        expect(
          () => validateSpec(spec),
          throwsA(_validationMessage(contains('unrecognized body syntax'))),
          reason: '$arrow $invalid',
        );
      }

      if (invalid.isNotEmpty) {
        final bare = parseSpec('Top::\n $invalid\n');
        expect(
          [
            for (final rule in bare.rules)
              for (final element in rule.body)
                element.kind is ActionEdgeBodyElementKind ||
                    element.kind is BlindEdgeBodyElementKind ||
                    element.kind is BareEdgeBodyElementKind,
          ],
          everyElement(isFalse),
          reason: 'bare $invalid',
        );
        expect(
          () => validateSpec(bare),
          throwsA(isA<SpecValidationException>()),
          reason: 'bare $invalid',
        );
      }
    }
  });

  test('invalid declarations never become suffix or prefix headers', () {
    for (final invalid in [
      '',
      'Top-Rule',
      'Top Rule',
      'Top😀',
      'Top:',
      'Top/Rule',
      r'$Top',
    ]) {
      expect(
        () => parseSpec('$invalid::\n /x/\n'),
        throwsA(isA<SpecParseException>()),
        reason: invalid,
      );
    }
  });

  test('validator rejects every programmatic declaration and target role', () {
    final invalidDeclaration = _specWith(
      label: 'Top-Rule',
      element: const RegexBodyElementKind(pattern: 'x'),
    );
    expect(
      () => validateSpec(invalidDeclaration),
      throwsA(_invalidLabel(role: 'declaration', label: 'Top-Rule', line: 1)),
    );

    final targetKinds = <BodyElementKind>[
      const ActionEdgeBodyElementKind(
        targets: [EdgeTarget(label: 'Bad-Target')],
      ),
      const BlindEdgeBodyElementKind(target: 'Bad-Target'),
      const BareEdgeBodyElementKind(
        targets: [BareEdgeTarget(label: 'Bad-Target')],
      ),
    ];
    for (final kind in targetKinds) {
      expect(
        () => validateSpec(_specWith(label: 'Top', element: kind)),
        throwsA(
          _invalidLabel(
            role: 'edge_target',
            label: 'Bad-Target',
            line: 2,
            owner: 'Top',
          ),
        ),
      );
    }
  });

  test('validator rejects invalid labels reconstructed from AST JSON', () {
    final encoded =
        jsonDecode(
              jsonEncode(
                _specWith(
                  label: 'Top',
                  element: const ActionEdgeBodyElementKind(
                    targets: [EdgeTarget(label: 'Child')],
                  ),
                ).toJson(),
              ),
            )
            as Map<String, Object?>;
    final rules = encoded['rules']! as List<Object?>;
    final owner = rules.single as Map<String, Object?>;
    final body = owner['body']! as List<Object?>;
    final element = body.single as Map<String, Object?>;
    final kind = element['kind']! as Map<String, Object?>;
    final targets = kind['targets']! as List<Object?>;
    final target = targets.single as Map<String, Object?>;
    target['label'] = 'Bad-Target';

    final reconstructed = SpecFile.fromJson(encoded);
    expect(
      () => validateSpec(reconstructed),
      throwsA(
        _invalidLabel(
          role: 'edge_target',
          label: 'Bad-Target',
          line: 2,
          owner: 'Top',
        ),
      ),
    );

    final header = owner['header']! as Map<String, Object?>;
    header['label'] = 'Bad-Declaration';
    final reconstructedInvalidDeclaration = SpecFile.fromJson(encoded);
    expect(
      () => validateSpec(reconstructedInvalidDeclaration),
      throwsA(
        _invalidLabel(role: 'declaration', label: 'Bad-Declaration', line: 1),
      ),
    );
  });
}

SpecFile _specWith({required String label, required BodyElementKind element}) {
  return SpecFile(
    rules: [
      Rule(
        header: RuleHeader(
          label: label,
          isTop: true,
          mode: RuleMode.defaultMode,
          rest: '',
          line: 1,
        ),
        body: [BodyElement(kind: element, source: '', line: 2)],
      ),
    ],
  );
}

Matcher _validationMessage(Matcher messageMatcher) {
  return isA<SpecValidationException>().having(
    (error) => error.message,
    'message',
    messageMatcher,
  );
}

Matcher _invalidLabel({
  required String role,
  required String label,
  required int line,
  String? owner,
}) {
  return isA<SpecValidationException>().having(
    (error) => error.diagnostic?.toJson(),
    'portable diagnostic',
    {
      'code': 'invalid_rule_label',
      'stage': 'validate_rule_labels',
      'message':
          "$role '$label' is not a nonempty Unicode 17.0.0 "
          'XID_Continue rule label',
      'fields': {
        'label': label,
        'line': line,
        'role': role,
        if (owner != null) 'rule_label': owner,
      },
    },
  );
}

File: julia/test/unicode_rule_label_routes_test.jl
# FUTURE-PARITY-BACKLOG.10.6.1.1 — Julia label parsing/validation routes.

function _julia_unicode_spec_with(label::AbstractString, kind)
    return SpecFile(rules = [
        Rule(
            header = RuleHeader(String(label), true, default_rule_mode(), "", 1),
            body = [BodyElement(kind, "", 2)],
        ),
    ])
end

function _julia_unicode_validation_error(spec::SpecFile)
    try
        validate_spec(spec)
    catch error
        return error
    end
    return nothing
end

function _julia_unicode_reconstruct(spec::SpecFile)
    encoded = JSON3.write(to_json(spec))
    return from_json(SpecFile, JSON3.read(encoded))
end

function _julia_unicode_invalid_label_json(
    role::AbstractString,
    label::AbstractString,
    line::Int;
    owner = nothing,
)
    fields = Dict{String,Any}(
        "label" => String(label),
        "line" => line,
        "role" => String(role),
    )
    if owner !== nothing
        fields["rule_label"] = String(owner)
    end
    return Dict{String,Any}(
        "code" => "invalid_rule_label",
        "stage" => "validate_rule_labels",
        "message" => "$(role) '$label' is not a nonempty Unicode 17.0.0 XID_Continue rule label",
        "fields" => fields,
    )
end

@testset "scanner parses every declaration and edge form with exact identity" begin
    decomposed_top = "To\u0308p"
    action = parse_spec("""
Töp::AND
 -> $decomposed_top | Δοκιμή { return(entry_text()) }

$decomposed_top:
 /x/

Δοκιμή:
 /x/
""")
    validate_spec(action)
    @test [rule.header.label for rule in action.rules] == ["Töp", decomposed_top, "Δοκιμή"]
    action_edge = only(action.rules[1].body).kind
    @test action_edge isa ActionEdgeBodyElementKind
    @test [target.label for target in action_edge.targets] == [decomposed_top, "Δοκιμή"]

    blind = parse_spec("""
規則::AND
 => 𐐀Rule

𐐀Rule:
 /x/
""")
    validate_spec(blind)
    blind_edge = only(blind.rules[1].body).kind
    @test blind_edge isa BlindEdgeBodyElementKind
    @test blind_edge.target == "𐐀Rule"

    bare = parse_spec("""
9_root::OR
 A·B | _ { return(entry_text()) }

A·B:
 /x/

_:
 /x/
""")
    validate_spec(bare)
    bare_edge = only(bare.rules[1].body).kind
    @test bare_edge isa BareEdgeBodyElementKind
    @test [target.label for target in bare_edge.targets] == ["A·B", "_"]
end

@testset "invalid suffixes never become partial action blind or bare edges" begin
    invalid_labels = ("", "Top-Rule", "Top Rule", "Top😀", "Top:", "Top/Rule", raw"$Top")
    for invalid in invalid_labels
        for arrow in ("->", "=>")
            spec = parse_spec(string("Top::\n ", arrow, " ", invalid, "\n"))
            @test only(only(spec.rules).body).kind isa RawBodyElementKind
            error = _julia_unicode_validation_error(spec)
            @test error isa SpecValidationException
            @test occursin("unrecognized body syntax", error.message)
        end

        if !isempty(invalid)
            spec = parse_spec(string("Top::\n ", invalid, "\n"))
            @test all(
                !(element.kind isa ActionEdgeBodyElementKind) &&
                !(element.kind isa BlindEdgeBodyElementKind) &&
                !(element.kind isa BareEdgeBodyElementKind)
                for rule in spec.rules for element in rule.body
            )
            @test _julia_unicode_validation_error(spec) isa SpecValidationException
        end
    end
end

@testset "invalid declarations never become suffix or prefix headers" begin
    invalid_labels = ("", "Top-Rule", "Top Rule", "Top😀", "Top:", "Top/Rule", raw"$Top")
    for invalid in invalid_labels
        @test_throws SpecParseException parse_spec(string(invalid, "::\n /x/\n"))
    end
    @test_throws SpecParseException parse_spec("Top:::\n /x/\n")
end

@testset "validator rejects every programmatic declaration and target role" begin
    invalid_declaration = _julia_unicode_spec_with("Top-Rule", RegexBodyElementKind("x"))
    declaration_error = _julia_unicode_validation_error(invalid_declaration)
    @test declaration_error isa SpecValidationException
    @test to_json(declaration_error.diagnostic) ==
        _julia_unicode_invalid_label_json("declaration", "Top-Rule", 1)

    target_kinds = (
        ActionEdgeBodyElementKind(targets = [EdgeTarget(label = "Bad-Target")]),
        BlindEdgeBodyElementKind(target = "Bad-Target"),
        BareEdgeBodyElementKind(targets = [BareEdgeTarget(label = "Bad-Target")]),
    )
    for kind in target_kinds
        error = _julia_unicode_validation_error(_julia_unicode_spec_with("Top", kind))
        @test error isa SpecValidationException
        @test to_json(error.diagnostic) ==
            _julia_unicode_invalid_label_json("edge_target", "Bad-Target", 2; owner = "Top")
    end
end

@testset "validator rejects invalid labels reconstructed from AST JSON" begin
    invalid_target = _julia_unicode_spec_with(
        "Top",
        ActionEdgeBodyElementKind(targets = [EdgeTarget(label = "Bad-Target")]),
    )
    target_error = _julia_unicode_validation_error(_julia_unicode_reconstruct(invalid_target))
    @test target_error isa SpecValidationException
    @test to_json(target_error.diagnostic) ==
        _julia_unicode_invalid_label_json("edge_target", "Bad-Target", 2; owner = "Top")

    invalid_declaration = _julia_unicode_spec_with("Bad-Declaration", RegexBodyElementKind("x"))
    declaration_error = _julia_unicode_validation_error(
        _julia_unicode_reconstruct(invalid_declaration),
    )
    @test declaration_error isa SpecValidationException
    @test to_json(declaration_error.diagnostic) ==
        _julia_unicode_invalid_label_json("declaration", "Bad-Declaration", 1)
end
